File: pdf_form_automator/fields.py
"""Candidate post-processing: unique field names, stripping existing fields."""

from __future__ import annotations

import fitz

from .heuristics import Candidate, slugify

TYPE_SUFFIX = {"date": "Datum", "signature": "Unterschrift", "checkbox": "Checkbox"}


def assign_names(candidates: list[Candidate]) -> None:
    used: set[str] = set()
    for cand in candidates:
        base = cand.name or slugify(cand.label)
        if base == "Feld" and cand.ftype in TYPE_SUFFIX:
            base = TYPE_SUFFIX[cand.ftype]
        name = base
        n = 2
        while name in used:
            name = f"{base}_{n}"
            n += 1
        used.add(name)
        cand.name = name


def strip_existing_fields(doc: fitz.Document) -> int:
    """Delete all existing form fields. Returns how many were removed."""
    removed = 0
    for page in doc:
        widgets = list(page.widgets())
        for w in widgets:
            page.delete_widget(w)
            removed += 1
    return removed

File: web/src/fields.ts
/** Port of fields.py: unique field names. (Stripping existing fields moved
 * to writer.ts, where the PDF object layer lives.) */

import { slugify } from "./heuristics";
import type { Candidate } from "./types";

export const TYPE_SUFFIX: Record<string, string> = {
  date: "Datum",
  signature: "Unterschrift",
  checkbox: "Checkbox",
};

export function assignNames(candidates: Candidate[]): void {
  const used = new Set<string>();
  for (const cand of candidates) {
    let base = cand.name || slugify(cand.label);
    if (base === "Feld" && cand.ftype in TYPE_SUFFIX) {
      base = TYPE_SUFFIX[cand.ftype];
    }
    let name = base;
    let n = 2;
    while (used.has(name)) {
      name = `${base}_${n}`;
      n++;
    }
    used.add(name);
    cand.name = name;
  }
}
